src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs

```diff
@@ -11,6 +11,493 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+  IL_0007:  ldarg.1
+  IL_0008:  call       ""void System.Console.WriteLine(int)""
+  IL_000d:  ret
+}");
+        }
+
+        [Fact]
+        public void InParamCtorOptionalArg()
+        {
+            var comp = CompileAndVerify(@"
+using System;
+class C
+{
+    public static void Main()
+    {
+        int x = 1;
+        new C(in x);
+    }
+
+    public C(in int x, int y = 0)
+    {
+        Console.WriteLine(x);
+        Console.WriteLine(y);
+    }
+}", expectedOutput: @"1
+0");
+            comp.VerifyIL("C.Main", @"
+{
+  // Code size       12 (0xc)
+  .maxstack  2
+  .locals init (int V_0) //x
+  IL_0000:  ldc.i4.1
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_0
+  IL_0004:  ldc.i4.0
+  IL_0005:  newobj     ""C..ctor(in int, int)""
+  IL_000a:  pop
+  IL_000b:  ret
+}");
+        }
+
+        [Fact]
+        public void InParamCollectionInitializerOptionalArg()
+        {
+            var comp = CompileAndVerify(@"
+using System;
+using System.Collections;
+class C : IEnumerable
+{
+    public static void Main()
+    {
+        int x = 1;
+        new C() { x };
+    }
+
+    public IEnumerator GetEnumerator() => null;
+
+    public void Add(in int x, int y = 0)
+    {
+        Console.WriteLine(x);
+        Console.WriteLine(y);
+    }
+}");
+        }
+
+        [Fact]
+        public void InParamSetter()
+        {
+            var comp = CompileAndVerify(@"
+using System;
+using System.Collections;
+class C
+{
+    static int _f = 1;
+    public static void Main()
+    {
+        new C()[in _f] = 0;
+    }
+
+    public IEnumerator GetEnumerator() => null;
+
+    public int this[in int x, int y = 0]
+    {
+        get => x;
+        set
+        {
+            Console.WriteLine(x);
+            Console.WriteLine(y);
+            _f++;
+            Console.WriteLine(x);
+        }
+    }
+}", expectedOutput: @"1
+0
+2");
+            comp.VerifyIL("C.Main", @"
+{
+  // Code size       18 (0x12)
+  .maxstack  4
+  IL_0000:  newobj     ""C..ctor()""
+  IL_0005:  ldsflda    ""int C._f""
+  IL_000a:  ldc.i4.0
+  IL_000b:  ldc.i4.0
+  IL_000c:  call       ""void C.this[in int, int].set""
+  IL_0011:  ret
+}");
+        }
+
+        [Fact]
+        public void InParamParamsArg()
+        {
+            var comp = CompileAndVerify(@"
+using System;
+class C
+{
+    public static void Main()
+    {
+        int x = 1;
+        M(in x);
+    }
+    public static void M(in int x, params int[] p)
+    {
+        Console.WriteLine(x);
+        Console.WriteLine(p.Length);
+    }
+}", expectedOutput: @"1
+0");
+            comp.VerifyIL("C.M", @"
+{
+  // Code size       16 (0x10)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  ldind.i4
+  IL_0002:  call       ""void System.Console.WriteLine(int)""
+  IL_0007:  ldarg.1
+  IL_0008:  ldlen
+  IL_0009:  conv.i4
+  IL_000a:  call       ""void System.Console.WriteLine(int)""
+  IL_000f:  ret
+}");
+        }
+
+        [Fact]
+        public void InParamReorder()
```

Comment:
Please add a test where there are three parameters that require a reorder while building for release. That will force us to spill at least two temporaries. Wanted to have a test that ensures the temporary optimizer doesn't incorrectly re-use a slot here.